src/ai.ts

```diff
@@ -1,32 +1,219 @@
+      }
+    }
+  }
+
+  if (jsonType === "start_object") fullMessage = "{" + fullMessage;
+  else if (jsonType === "start_array") fullMessage = "[" + fullMessage;
+  if (jsonType === "start_object" || jsonType === "start_array")
+    fullMessage = fullMessage.split("```")[0];
+  if (jsonType === "parse") {
+    const matchedJSON = fullMessage.match(/```json([\s\S]*?)```/g);
+    if (matchedJSON) {
+      fullMessage = matchedJSON[0]
+        .replace(/```json/g, "")
+        .replace(/```/g, "")
+        .trim();
+    }
+  }
+
+  return {
+    fullMessage,
+    outputTokens,
+    inputTokens
+  };
+}
+
+// TODO:
+// * Implement streaming. Doesn't look like OpenAI includes token usage in stream response object though, so that becomes complicated.
+// * `json_mode` doesn't really support continuance since it always returns proper JSON objects. Need to figure out how to handle that - json mode doesn't guarantee that it will actually finish the JSON object.
+async function callOpenAI(
+  messages: GenericMessageParam[],
+  options: AICallerOptions
+): Promise<AICallResponse> {
+  const {
+    model,
+    maxOutputTokens,
+    streamToConsole,
+    saveToFilepath,
+    apiKey,
+    prefix,
+    systemPrompt,
+    jsonType
+  } = options;
+  console.log("Calling OpenAI");
+
+  const arrayWrapperPromptAddition =
+    jsonType === "start_array"
+      ? "Wrap the array in an object with the key `response`."
+      : "";
+
+  if (systemPrompt) {
+    messages.unshift({
+      role: "system",
+      content: systemPrompt + arrayWrapperPromptAddition
+    });
+  }
+
+  const openai = apiKey ? new OpenAI({ apiKey }) : new OpenAI();
+
+  const completion = await openai.chat.completions.create({
+    messages,
+    model,
+    max_tokens: maxOutputTokens,
+    response_format: jsonType ? { type: "json_object" } : undefined
+  });
+  let fullMessage = completion.choices[0].message.content || "";
+  const inputTokens = completion.usage?.prompt_tokens || 0; // I have no idea why usage can be undefined???
+  const outputTokens = completion.usage?.completion_tokens || 0;
+
+  if (streamToConsole) {
+    process.stdout.write(
+      `\n\nStreaming from ${model}${
+        saveToFilepath ? ` to ${saveToFilepath}` : ""
+      }: `
+    );
+    process.stdout.write(fullMessage);
+  }
+  if (saveToFilepath) {
+    fs.writeFileSync(saveToFilepath, (prefix || "") + fullMessage);
+  }
+
+  if (jsonType === "start_array") {
+    fullMessage = JSON.stringify(JSON.parse(fullMessage).response);
+  }
+
+  return { fullMessage, inputTokens, outputTokens };
+}
+
+// note: should we call this caLLM for fun?
+export async function callLLM(
+  messages: GenericMessageParam[],
+  options: AICallerOptions
+): Promise<AICallSuccess | AICallFailure> {
+  const {
+    model,
+    maxOutputTokens,
+    apiKey,
+    streamToConsole = false, // Set default values here
+    saveName,
+    jsonType,
+    saveToFilepath,
+    prefix,
+    systemPrompt,
```

Comment:
@HebeHH this seems unused?

Reply:
hmm it's used when the `options` are passed to the specific callProvider() functions, but I guess that doesn't mean they need to be pulled out here. 
Personally I kinda like being able to see what the parameters available to the function are, for future dev even if not in use right now, but it can also freely be removed.

Reply:
I think typescript means that you can see them from the typedef anyway :)

Reply:
yup but the typedef is a whole other file. However you can see it on hovering the typedef in VS Code so I shall remove